nums/core/array/utils.py

```diff
@@ -26,6 +26,46 @@
 # pylint: disable = no-member, trailing-whitespace
 
 
+def find_diag_output_blocks(X_blocks, total_elements):
```

Comment:
Can you nest this function inside of `diag` in `ArrayApplication`? I did not realize it's operating on the blocks of X directly, and `utils` thus far does not operate on blocks directly. Also, let's change the signature of this function to `X: BlockArray, total_elements: int`. Within the body of the code, just use `X.blocks` instead of `X_blocks`.